### ISSUE TITLE
Release 1.7.0

### ISSUE THREAD

**Issue by pepeiborra** [haskell/haskell-language-server#2739]
The main change in this release will be the switch to dynamically linked binaries for distribution, contributed by @hasufell 

As usual we need a volunteer to drive the release.

## Release blockers

- [ ] VSCode extension upgraded to use ghcup for installation

## Prerelease steps

- [ ] trigger manually the hackage workflow without uploading the packages
- [ ] trigger manually the build workflow
- [ ] create a prerelease tag 1.7.0-check-gitlab and push it to the [project repo in gitlab](https://gitlab.haskell.org/haskell/haskell-language-server) to check the build is fine

## Github release steps

- [ ] generate the list of pull requests finished since the last release using the [haskell script](https://github.com/haskell/haskell-language-server/blob/master/GenChangelogs.hs) in the project root.
  Nix users should run command `gen-hls-changelogs` (a wrapper of the script) in nix-shell instead.
- [ ] add that list to the actual [Changelog](https://github.com/haskell/haskell-language-server/blob/master/ChangeLog.md) with a description of the release.
- [ ] bump up versions of changed packages. All are optional but [haskell-language-server itself](https://github.com/haskell/haskell-language-server/blob/master/haskell-language-server.cabal).
- [ ] create the tag and make an initial prerelease to trigger the ci workflow (see details below)
- [ ] contact ghcup team (#haskell-ghcup irc channel or via its [repo](https://github.com/haskell/ghcup-metadata)) to try to sync our release and its inclusion in ghcup
- [ ] in the github release edit page, check the attached binaries and the release description (usually the changelog entry) and uncheck the prerelease box
- [ ] make public the release in the usual social channels (not required but useful to spread the word :slightly_smiling_face:):
  - [ ] irc
  - [ ] matrix
  - [ ] twitter
  - [ ] discord
  - [ ] discourse
  - [ ] reddit
  - [ ] 
### hackage release

- [ ] bump up package versions following the [pvp specification](https://pvp.haskell.org/) if they are not already updated. You *must* use [policeman](https://github.com/kowainik/policeman) (build from HEAD) to help with this step.
- [ ] create 1.7.0-hackage branch to trigger the hackage github workflow which will upload all changed packages to hackage as candidates
- [ ] check manually candidates in hackage
- [ ] publish them definitely

### ghcup release

- [ ] push the release tag to the [haskell-language-server gitlab repo](https://gitlab.haskell.org/haskell/haskell-language-server) to trigger the build of ghcup specific artifacts
- [ ] download specific artifacts only available in the gitlab build and compute their sha256sum
- [ ] upload them to the github release and complete the SHA256SUMS file
- [ ] change ghcup metadata to include the new release in https://github.com/haskell/ghcup-metadata
  -  example pull request [here](https://github.com/haskell/ghcup-metadata/pull/19)

**Comment by hasufell** [comment]
I'm preparing a ghcup release.

**Comment by hasufell** [comment]
Released.

**Comment by hasufell** [comment]
This should also be merged before release https://github.com/haskell/haskell-language-server/pull/2738

Who is maintaning the vscode extension?

**Comment by fendor** [comment]
> Who is maintaning the vscode extension?

@jneira takes care of releases, iirc

**Comment by jneira** [comment]
Doing the release is matter of push a tag, create a github release and make it definitive. 
Feel free to ask any question about when doing it. 
Many of hls maintainers are also maintainers of the vscode extension and can do merges, reviews and releases but let me know if anyone needs some type of permission to do the required changes.

**Comment by hasufell** [comment]
The question is how does vscode extension install HLS?

**Comment by jneira** [comment]
Here is the main entry point: https://github.com/haskell/vscode-haskell/blob/79456aaaa98b98f9e7b9f690bb18ab2b2838b6e2/src/extension.ts#L194

**Comment by hasufell** [comment]
It appears not: https://github.com/haskell/vscode-haskell/blob/79456aaaa98b98f9e7b9f690bb18ab2b2838b6e2/src/hlsBinaries.ts#L317

**Comment by pepeiborra** [comment]
One issue with using ghcup to install HLS is specifying the target GHC version, which can be custom to the cradle and different from the one set in ghcup or the one in the PATH. Currently the VSCode extension uses the hls-wrapper to find out the project GHC version.

If not using ghcup, we would need to teach the VSCode extension to use the make script included in the new HLS binary distribution

**Comment by hasufell** [comment]
> One issue with using ghcup to install HLS is specifying the target GHC version, which can be custom to the cradle and different from the one set in ghcup or the one in the PATH. Currently the VSCode extension uses the hls-wrapper to find out the project GHC version.
> 
> If not using ghcup, we would need to teach the VSCode extension to use the make script included in the new HLS binary distribution

What's the exact issue? ghcup just installs all of them.

**Comment by pepeiborra** [comment]
Validation: ghcup doesn't know what the project ghc is so it cannot validate whether HLS binaries exist.

**Comment by hasufell** [comment]
> Validation: ghcup doesn't know what the project ghc is so it cannot validate whether HLS binaries exist. 

Why would ghcup need to know that? The VSCode extension knows and can easily check whether a `haskell-language-server-<ghcver>` binary exists.

**Comment by pepeiborra** [comment]
I thought that we wanted to make the VSCode extension dumb and delegate all the work on GHCup.

**Comment by hasufell** [comment]
> I thought that we wanted to make the VSCode extension dumb and delegate all the work on GHCup.

I don't see how that would work. GHCup isn't going to add functionality specific to a VSCode extension. It can install HLS into an isolated dir easily without affecting the global installation.

So something along these lines:

1. make ghcup available
    * if it already exists in `releasesDownloadStoragePath`, run `<ghcup-binary> upgrade --inplace`
    * if it doesn't exist, download it from `https://downloads.haskell.org/~ghcup/<arch>-<platform>-ghcup`
2. check if HLS is in `<releasesDownloadStoragePath>/bin`
    * if it already exists and is the most recent version, do nothing
    * if it exists, but is outdated remove `<releasesDownloadStoragePath>/bin` and `<releasesDownloadStoragePath>/lib`, then run `<ghcup-binary> install hls --isolate <releasesDownloadStoragePath> --force <hls-ver>`
    * if it doesn't exist, run `<ghcup-binary> install hls --isolate <releasesDownloadStoragePath> --force <hls-ver>`
3. Check if `<releasesDownloadStoragePath>/bin/haskell-language-server-<project-ghcver>` exists
    * if yes, succeed
    * if not, abort and tell the user to compile HLS from source

So, is anyone going to work on the VSCode extension? I'm not a VSCode user and not familiar with the code.

**Comment by hasufell** [comment]
https://github.com/haskell/vscode-haskell/pull/533

**Comment by fendor** [comment]
@wz1000 reported some bug with the new hie-bios version

**Comment by hasufell** [comment]
Another release blocker: https://github.com/haskell/haskell-language-server/pull/2746

**Comment by hasufell** [comment]
> @wz1000 reported some bug with the new hie-bios version. No details yet

Do we have details now?

**Comment by fendor** [comment]
One issue:

ghc-pkg wrapper script is generated in a wrong way:
```sh
exec ./autogenerated_hoogle_database_path\ ::\ Data.String.IsString\ p\ =>\ p --global-package-db /home/zubin/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/package.conf.d ${1+"$@"}
```

**Comment by wz1000** [comment]
I can do the HLS release this time, but I will need hackage permissions for the many packages.

**Comment by Abastro** [comment]
Is the hie-bios update included here? I am having problem with HLS not recognizing "with-compiler" option.

**Comment by pepeiborra** [comment]
> I can do the HLS release this time, but I will need hackage permissions for the many packages.

Thank you !!! I will created an `hls_team` umbrella Hackage account and will share auth details via IRC

**Comment by fendor** [comment]
@Abastro What hie-bios update? This release comes with hie-bios 0.9.0, the latest hackage release and without new patches.

Can you specify our issue?

**Comment by Abastro** [comment]
> @Abastro What hie-bios update? This release comes with hie-bios 0.9.0, the latest hackage release and without new patches.
> 
> Can you specify our issue?

Yes, hie-bios 0.9.0 comes with a fix to conform to cabal's "with-compiler".
I am looking forward to new hls release coming with hie-bios 0.9.0, it seems like I cannot install the 0.9.0 version along with currently available hls.

**Comment by fendor** [comment]
@Abastro Why not? I installed it from source, and it worked out of the box. Or do you mean you tried installing hie-bios alongside HLS? You'll have to build HLS from source, too.

**Comment by fendor** [comment]
Installation failed, the feature did not work or HLS crashed?

**Comment by Abastro** [comment]
Did not install. Cabal does not even try hie-bios 0.9.0., even with `allow-newer`.

**Comment by Abastro** [comment]
Yep, also fails - with following cabal part log.
```
[ cabal ] hie-bios-0.2.0, hie-bios-0.1.1, hie-bios-0.1.0, hie-bios-0.7.3 (constraint
[ cabal ] from user target requires ==0.9.0)
[ cabal ] [__1] fail (backjumping, conflict set: ghcide, hie-bios)
[ cabal ] After searching the rest of the dependency tree exhaustively, these were the
[ cabal ] goals I've had most trouble fulfilling: hie-bios, ghcide
[ cabal ]
```
The command I tried: `ghcup compile hls -v 1.6.1.0 --ghc 8.10.7 --ghc 9.0.2 --cabal-project-local https://gist.githubusercontent.com/hasufell/8409effe97d33c36369e4d9906556c1e/raw/0d6caaabaa974430c2c3b3ba68a4872d2a984dbf/gistfile1.txt`

**Comment by hasufell** [comment]
> ghcup compile hls -v 1.6.1.0 --ghc 8.10.7 --ghc 9.0.2 --cabal-project-local https://gist.githubusercontent.com/hasufell/8409effe97d33c36369e4d9906556c1e/raw/0d6caaabaa974430c2c3b3ba68a4872d2a984dbf/gistfile1.txt

```
ghcup compile hls -g master --ghc 8.10.7 --cabal-project-local https://gist.githubusercontent.com/hasufell/8409effe97d33c36369e4d9906556c1e/raw/0
d6caaabaa974430c2c3b3ba68a4872d2a984dbf/gistfile1.txt
```

You can't build 8.10.7 and 9.0.2 in one go, because 9.0.2 needs `--cabal-project cabal-ghc90.project`.

**Comment by Abastro** [comment]
Oh, I see. Thank you!

**Comment by Abastro** [comment]
Seems like installing the master branch did the trick. Anyway, thanks! (Still, would be grateful for the formal release)

**Comment by hasufell** [comment]
who has an overview of that GHCs are to be supported?

These two files will need adjustment:

1. https://github.com/haskell/haskell-language-server/blob/master/bindist/ghcs
2. https://github.com/haskell/haskell-language-server/blob/master/bindist/ghcs-FreeBSD

**Comment by pepeiborra** [comment]
Good point, we'll need to add support for GHC 9.2.2 before the 1.7.0 release.

**Comment by hasufell** [comment]
> Good point, we'll need to add support for GHC 9.2.2 before the 1.7.0 release.

9.2.2 is [broken on windows](https://gitlab.haskell.org/ghc/ghc/-/issues/21196) ... not sure it's worth the support.

**Comment by Abastro** [comment]
How about GHC 9.2.1?

**Comment by hasufell** [comment]
9.2.1 has severe bugs in native M1 code generator. I'm not totally sure if 8.10.7 with llvm backend is better, though.

https://downloads.haskell.org/~ghc/9.2.2/docs/html/users_guide/9.2.2-notes.html gives some.overview

**Comment by georgefst** [comment]
As discussed in https://github.com/haskell/haskell-language-server/pull/2763#issuecomment-1066140768, the functionality added in that PR is possibly in a weird state, of being implemented but not supported by editor integrations. It would be a shame to release it like that, so it would be great to get some help from someone who knows more about how plugin-specific options work.

**Comment by michaelpj** [comment]
I think we can do that after releasing the server if we need to. Having the option isn't harmful by itself.

**Comment by jrp2014** [comment]
Does this release support Apple M1 Macs natively (ie, not just through LLVM). Sorry if this has been answered multiple times before.

**Comment by hasufell** [comment]
I suggest to move forward with the release and not let the vscode extension block it.

The only decision that needs to be made for that is whether we still distribute the static binaries on the github release page. I'd vote for no (especially since that logic is removed from the github workflow). The extension currently would ignore the newly uploaded make-based tarballs afais: https://github.com/haskell/vscode-haskell/blob/70f6883af669ef176fe3cf8ec4963921de686f79/src/hlsBinaries.ts#L382-L413

So we can just go ahead.

**Comment by pepeiborra** [comment]
> I can do the HLS release this time, but I will need hackage permissions for the many packages.

@wz1000 are you still good to do the 1.7.0 release?

**Comment by wz1000** [comment]
Yes, I will do the release this week.

**Comment by wz1000** [comment]
All the binary and hackage releases have been made.
